src/pwm.rs

```diff
@@ -12,98 +12,136 @@ use core::time::Duration;
+/// #     const MAX_DUTY: u16 = 1024;
 /// #     type Error = Infallible;
-/// #     type Duty = u16;
+/// #
 /// #     fn try_disable(&mut self, _: usize) -> Result<(), Self::Error> { unimplemented!() }
 /// #     fn try_enable(&mut self, _: usize) -> Result<(), Self::Error> { unimplemented!() }
-/// #     fn try_get_duty(&self, _: usize) -> Result<u16, Self::Error> { unimplemented!() }
-/// #     fn try_get_max_duty(&self) -> Result<u16, Self::Error> { Ok(0) }
-/// #     fn try_set_duty(&mut self, _: usize, _: u16) -> Result<(), Self::Error> { Ok(()) }
 /// #     fn try_get_period(&self) -> Result<Duration, Self::Error> { unimplemented!() }
 /// #     fn try_set_period<T>(&mut self, _: T) -> Result<(), Self::Error> where T: Into<Duration> { Ok(()) }
 /// # }
+/// #
+/// # impl hal::pwm::PwmDuty<u16> for Pwm1<HasPins> {
+/// #     fn try_get_duty(&self, _: usize) -> Result<u16, Self::Error> { unimplemented!() }
+/// #     fn try_set_duty<D>(&mut self, _: usize, _: D) -> Result<(), Self::Error> where D: Into<u16> { Ok(()) }
+/// # }
+/// # impl hal::pwm::PwmPin<u16> for Pwm1Pin {
+/// #     const MAX_DUTY: u16 = 1024;
+/// #     type Error = Infallible;
+/// #     fn try_disable(&mut self) -> Result<(), Self::Error> { unimplemented!() }
+/// #     fn try_enable(&mut self) -> Result<(), Self::Error> { unimplemented!() }
+/// #     fn try_get_duty(&self) -> Result<u16, Self::Error> { unimplemented!() }
+/// #     fn try_set_duty<D>(&mut self, _: D) -> Result<(), Self::Error> where D: Into<u16> { Ok(()) }
+/// # }
 /// ```
 // unproven reason: pre-singletons API. The `PwmPin` trait seems more useful because it models independent
 // PWM channels. Here a certain number of channels are multiplexed in a single implementer.
-pub trait Pwm {
+pub trait Pwm<Duty> {
     /// Number of available channels
     const CHANNELS: usize;
 
+    /// Maximum duty cycle value
+    const MAX_DUTY: Duty;
+
     /// Enumeration of `Pwm` errors
     type Error;
 
-    /// Type for the `duty` methods
-    ///
-    /// The implementer is free to choose a float / percentage representation
-    /// (e.g. `0.0 .. 1.0`) or an integer representation (e.g. `0 .. 65535`)
-    type Duty;
-
     /// Disables a PWM `channel`
     fn try_disable(&mut self, channel: usize) -> Result<(), Self::Error>;
 
     /// Enables a PWM `channel`
     fn try_enable(&mut self, channel: usize) -> Result<(), Self::Error>;
 
+    // Note: should you be able to _set_ the period once the channels have been split?
+    // my feeling is, probably not
+
     /// Returns the current PWM period
     fn try_get_period(&self) -> Result<Duration, Self::Error>;
 
+    /// Sets a new PWM period
+    fn try_set_period<P>(&mut self, period: P) -> Result<(), Self::Error>
+    where
+        P: Into<Duration>;
+
+    // Note: technically there could be a `channel` or `split` method here but this is
+    // rather extremely difficult prior to const-generics landing (and CHANNELS would need to be a const generic)
```

Comment:
I think we shouldn't concern ourselves with `split` here. Initialization and configuration have always been out of scope for embedded-hal, and I think that's for the better. Just provide I/O interfaces that can be provided by HALs and consumed by drivers. Setting those up is the user's responsibility, and probably to device-specific to be abstracted over in a zero-cost/unopinionated way.

Reply:
yeah this is fair, we do have to communicate the expectations around split'ing to hal implementers tho, so if there's any opportunity for us to smooth that experience it seemed worth a look? can't do till const-generics and mostly agreed about not being opinionated about this just, exploring options.

Reply:
I can see a future where the design space is generally much better understood, and where adding such constraints makes sense. I don't think we're there yet.

+1 to exploring options though!